compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/HashFunction.java

```diff
@@ -0,0 +1,170 @@
+
+    /**
+     * Generates LIR that implements the hash function in terms of value and min.
+     *
+     * @param value the value to be hashed
+     * @param min the lowest key
+     * @param gen the lir generator
+     * @return new lir value with the hash function applied
+     */
+    public abstract Value gen(Value value, Value min, ArithmeticLIRGenerator gen);
+
+    /**
+     * Returns an estimate of number of CPU cycles necessary to apply the hash function.
+     */
+    public abstract int effort();
+
+    /**
+     * @return a list of all available hash functions
+     */
+    public static final List<HashFunction> instances() {
+        return Collections.unmodifiableList(instances);
+    }
+
+    private static List<HashFunction> instances = new ArrayList<>();
+
+    private static int[] mersennePrimes = {3, 7, 31, 127, 8191, 131071, 524287, 2147483647};
+
+    static {
+      //@formatter:off
+
+        add("val", 0,
+            (val, min) -> val,
+            gen -> (val, min) -> val);
+
+        add("val - min", 1,
+            (val, min) -> val - min,
+            gen -> (val, min) -> gen.emitSub(val, min, false));
+
+        add("val >> min", 1,
+            (val, min) -> val >> min,
+            gen -> (val, min) -> gen.emitShr(val, min));
+
+        add("val >> (val & min)", 2,
+            (val, min) -> val >> (val & min),
+            gen -> (val, min) -> gen.emitShr(val, gen.emitAnd(val, min)));
+
+        add("(val >> min) ^ val", 2,
+            (val, min) -> (val >> min) ^ val,
+            gen -> (val, min) -> gen.emitXor(gen.emitShr(val, min), val));
+
+        add("(val >> min) * val", 3,
+            (val, min) -> (val >> min) * val,
+            gen -> (val, min) -> gen.emitMul(gen.emitShr(val, min), val, false));
+
+        addWithPrimes("(val * prime) >> min", 3,
+                      prime -> (val, min) -> (val * prime) >> min,
+                      (gen, prime) -> (val, min) -> gen.emitShr(gen.emitMul(val, prime, false), min));
+
+        addWithPrimes("rotateRight(val, prime)", 3,
+                      prime -> (val, min) -> Long.rotateRight(val, prime),
+                      (gen, prime) -> (val, min) -> rotateRight(gen, val, prime));
+
+        addWithPrimes("rotateRight(val, prime) + val", 4,
+                      prime -> (val, min) -> Long.rotateRight(val, prime) + val,
+                      (gen, prime) -> (val, min) -> gen.emitAdd(rotateRight(gen, val, prime), val, false));
+
+        addWithPrimes("rotateRight(val, prime) ^ val", 4,
+                      prime -> (val, min) -> Long.rotateRight(val, prime) ^ val,
+                      (gen, prime) -> (val, min) -> gen.emitXor(rotateRight(gen, val, prime), val));
+      //@formatter:on
+    }
+
+    private static Value rotateRight(ArithmeticLIRGenerator gen, Value i, Value distance) {
+        // (i >>> distance) | (i << -distance)
+        return gen.emitOr(gen.emitUShr(i, distance), gen.emitShl(i, gen.emitNegate(distance)));
```

Comment:
AMD64ArithmeticLIRGenerator provides emitRor which would be a lot more efficient.  Maybe provide a platform independent definition like above and use emitRor for AMD64?

Reply:
@tkrodriguez maybe I could add `emitRor` to `ArithmeticLIRGenerator` with the slow implementation as the default and then `AMD64ArithmeticLIRGenerator` overrides it with a better impl?